docs/k2hb-configurations.md
# Configuration

There are a number of environment variables that can be used to configure
the system. Some of them are for configuring Kafka2Hbase itself, and some
are for configuring the built-in ACM PCA client to perform mutual auth.

## Kafka2Hbase Configuration

#### Hbase

By default Kafka2Hbase will connect to Zookeeper at `zookeeper:2181` use the parent uri `hbase`
and create tables in the `k2hb` namespace. The data will be stored in `cf:data`
with at least `1` version and at most `10` versions and a TTL of 10 days.

* **K2HB_HBASE_ZOOKEEPER_PARENT**
    The hbase parent uri, defaults to `/hbase` but should be set to `/hbase-unsecure` for AWS HBase
* **K2HB_HBASE_ZOOKEEPER_QUORUM**
    Comma separated list of Zookeeper servers
* **K2HB_HBASE_ZOOKEEPER_PORT**
    The listening port of the Zookeeper servers
* **K2HB_HBASE_DATA_TABLE**
    The name of the table to store message bodies in
* **K2HB_HBASE_DATA_FAMILY**
    The name of the column family to store message bodies in
* **K2HB_HBASE_TOPIC_TABLE**
    The name of the table to store topic message counts in
* **K2HB_HBASE_TOPIC_FAMILY**
    The name of the column family to store topic message counts in
* **K2HB_HBASE_TOPIC_QUALIFIER**
    The name of the column qualifier to store topic message counts in

#### Kafka

By default Kafka2Hbase will connect to Kafka at `kafka:9092` in the `k2hb`
consumer group. It will poll the `test-topic` topic with a poll timeout of
`10` days, and refresh the topics list every 10 seconds (`10000` ms).

* **K2HB_KAFKA_BOOTSTRAP_SERVERS**
    Comma separated list of Kafka servers and ports
* **K2HB_KAFKA_CONSUMER_GROUP**
    The name of the consumer group to join
* **K2HB_KAFKA_TOPIC_REGEX**
    A regex that will fetch a list of topics to listen to, e.g. `db.*`. Defaults to `test-topic.*`
* **K2HB_KAFKA_META_REFRESH_MS** (Optional)
    The frequency that the consumer will ask the broker for metadata updates, which also checks for new topics.
    Defaults to `10000` ms (10 seconds).
    Typically, should be an order of magnitude less than `K2HB_KAFKA_POLL_TIMEOUT`, else new topics will not be discovered within each polling interval.
* **K2HB_KAFKA_POLL_TIMEOUT**
    The maximum time to wait for messages in ISO-8601 duration format (e.g. `PT10S`).
    Defaults to 1 Hour.
    Should be greater than `K2HB_KAFKA_META_REFRESH_MS`, else new topics will not be discovered within each polling interval.
* **K2HB_KAFKA_INSECURE**
    Disable SSL entirely (useful for dev / test) with `K2HB_KAFKA_INSECURE=true`
* **K2HB_KAFKA_CERT_MODE**
    If SSL is enabled, either create certs in ACM-PCA with value `CERTGEN` or retrieve
    them from ACM with value `RETRIEVE`

#### SSL Mutual Authentication (CERTGEN mode)

By default the SSL is enabled but has no defaults. These must either be
configured in full or disabled entirely via `K2HB_KAFKA_INSECURE=FALSE`
and `K2HB_KAFKA_CERT_MODE=CERTGEN`.

For an authoritative full list of arguments see the tool help; Arguments not listed here are
defaulted in the `entrypoint.sh` script.

* **CERTGEN_CA_ARN**
    The AWS CA ARN to use to generate the cert
* **CERTGEN_KEY_TYPE**
    The type of private key (`RSA` or `DSA`)
* **CERTGEN_KEY_LENGTH**
    The key length in bits (`1024`, `2048` or `4096`)
* **CERTGEN_KEY_DIGEST**
    The key digest algorithm (`sha256`, `sha384`, `sha512`)
* **CERTGEN_SUBJECT_C**
    The subject country
* **CERTGEN_SUBJECT_ST**
    The subject state/province/county
* **CERTGEN_SUBJECT_L**
    The subject locality
* **CERTGEN_SUBJECT_O**
    The subject organisation
* **CERTGEN_SUBJECT_OU**
    The subject organisational unit
* **CERTGEN_SUBJECT_EMAILADDRESS**
    The subject email address
* **CERTGEN_SIGNING_ALGORITHM**
    The certificate signing algorithm used by ACM PCA
    (`SHA256WITHECDSA`, `SHA384WITHECDSA`, `SHA512WITHECDSA`, `SHA256WITHRSA`, `SHA384WITHRSA`, `SHA512WITHRSA`)
* **CERTGEN_VALIDITY_PERIOD**
    The certificate validity period in Go style duration (e.g. `1y2m6d`)
* **CERTGEN_PRIVATE_KEY_ALIAS**
    Alias for the private key
* **CERTGEN_TRUSTSTORE_CERTS**
    Comma delimited list of S3 URIs pointing to certificates to be included in the trust store
* **CERTGEN_TRUSTSTORE_ALIASES**
    Comma delimited list of aliases for the certificate
* **CERTGEN_LOG_LEVEL**
    The log level of the certificate generator (`CRITICAL`, `ERROR`, `WARNING`, `INFO`, `DEBUG`)


#### SSL Mutual Authentication (RETRIEVE mode)

By default the SSL is enabled but has no defaults. These must either be
configured in full or disabled entirely via `K2HB_KAFKA_INSECURE=FALSE`
and `K2HB_KAFKA_CERT_MODE=RETRIEVE`.

For an authoritative full list of arguments see the tool help; Arguments not listed here are
defaulted in the `entrypoint.sh` script.

* **RETRIEVER_ACM_CERT_ARN**
    ARN in AWS ACM to use to fetch the required cert, cert chain, and key
* **RETRIEVER_ADD_DOWNLOADED_CHAIN**
    Whether or not to add the downloaded cert chain from the ARN to the trust store
    Allowed missing, `true`, `false`, `yes`, `no`, `1` or `0`
    If missing defaults to false
* **RETRIEVE_TRUSTSTORE_CERTS**
    Comma delimited list of S3 URIs pointing to certificates to be included in the trust store
* **RETRIEVE_TRUSTSTORE_ALIASES**
    Comma delimited list of aliases for the certificate
* **RETRIEVE_LOG_LEVEL**
    The log level of the certificate generator (`CRITICAL`, `ERROR`, `WARNING`, `INFO`, `DEBUG`)

#### Metadatastore 

| Name | Notes |
|------|--------|
|K2HB_METADATA_STORE_TABLE| The name of the table to write metadata store entries to |
|K2HB_RDS_CA_CERT_PATH|Where the Amazon root CA is located|
|K2HB_RDS_DATABASE_NAME|The name of the database to connect to ('metadatastore for circle/local)|
|K2HB_RDS_ENDPOINT|The server to connect to ('metadatastore for circle/local)|
|K2HB_RDS_PASSWORD_SECRET_NAME|The name of the password secret to look up (should be 'password' for local/circle running)|
|K2HB_RDS_PORT|The port to connect on ('3306' for local and circle)|
|K2HB_RDS_USERNAME|The username to connect with ('k2hbwriter' for local usage and circle)|
|K2HB_USE_AWS_SECRETS|Whether to fetch metadatastore passwords from AWS (should be 'false' for local running and circle)|
|K2HB_USE_AWS_SECRETS|Whether to look up secrets in AWS ('false' for local/circle)|
|K2HB_WRITE_TO_METADATA_STORE| Whether to write to the metadata store - to enable us to feature toggle writes on and off |

#### AWS Service
| Name | Notes|
|------|------|
|K2HB_AWS_S3_ARCHIVE_BUCKET| The bucket to which the messages are written |
|K2HB_AWS_S3_ARCHIVE_DIRECTORY| The directory in the bucket under which messages are written (or the common part of the key/prefix) |
|K2HB_AWS_S3_BATCH_PUTS| Whether to write each batch of received messages for a topic/partition combination as one object rather than one object per message, should probably be 'true' for actual running in aws |
|K2HB_AWS_S3_MAX_CONNECTIONS| Default max concurrentl connections is 500, set this to deviate from that|
|K2HB_AWS_S3_PARALLEL_PUTS| If not batch putting allows all messages to be written in parallel - n.b. may cause rate limiting issues |
|K2HB_AWS_S3_REGION| Set this if the default is not where you want to point |
|K2HB_AWS_S3_USE_LOCALSTACK| Set to true for local running - uses containerized aws|


src/main/kotlin/Config.kt

import com.amazonaws.ClientConfiguration
import com.amazonaws.Protocol
import com.amazonaws.auth.AWSStaticCredentialsProvider
import com.amazonaws.auth.BasicAWSCredentials
import com.amazonaws.auth.DefaultAWSCredentialsProviderChain
import com.amazonaws.client.builder.AwsClientBuilder
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.AmazonS3ClientBuilder
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.HConstants
import org.apache.kafka.clients.consumer.ConsumerConfig
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.common.config.SslConfigs
import org.apache.kafka.common.serialization.ByteArrayDeserializer
import org.apache.kafka.common.serialization.ByteArraySerializer
import java.io.File
import java.net.InetAddress
import java.time.Duration
import java.util.*
import kotlin.time.ExperimentalTime
import kotlin.time.minutes
import kotlin.time.seconds

@ExperimentalTime
object Config {
    const val metaDataRefreshKey = "metadata.max.age.ms"
    const val schemaFileProperty = "schema.location"
    const val mainSchemaFile = "business_message.schema.json"
    const val equalitySchemaFile = "equality_message.schema.json"
    const val auditSchemaFile = "audit_message.schema.json"
    const val dataworksRegion = "eu-west-2"

    object Metrics {
        val scrapeInterval: Long = getEnv("K2HB_METRICS_SCRAPE_INTERVAL")?.toLong() ?: 1.minutes.toLongMilliseconds()
        val deleteMetrics = getEnv("K2HB_METRICS_DELETE_METRICS")?.toBoolean() ?: true
        val instanceName = getEnv("K2HB_INSTANCE_NAME") ?: ""
        val pushgateway = getEnv("K2HB_METRICS_PUSHGATEWAY") ?: "pushgateway:9091"
        val pushMetrics = getEnv("K2HB_METRICS_PUSH_METRICS")?.toBoolean() ?: true

        val pushScheduleInitialDelay = getEnv("K2HB_METRICS_SCHEDULE_INITIAL_DELAY_SECONDS")?.run {
            toInt().seconds.inMilliseconds.toLong()
        } ?: 10.seconds.inMilliseconds.toLong()

        val pushSchedulePeriod = getEnv("K2HB_METRICS_SCHEDULE_PERIOD_SECONDS")?.run {
            toInt().seconds.inMilliseconds.toLong()
        } ?: 20.seconds.inMilliseconds.toLong()

    }

    object Shovel {
        val reportFrequency = getEnv("K2HB_KAFKA_REPORT_FREQUENCY")?.toInt() ?: 100
    }

    object Validator {
        var properties = Properties().apply {
            put(schemaFileProperty, getEnv("K2HB_VALIDATOR_SCHEMA") ?: mainSchemaFile)
        }
    }

    object Hbase {
        val config = Configuration().apply {
            set(HConstants.ZOOKEEPER_ZNODE_PARENT, getEnv("K2HB_HBASE_ZOOKEEPER_PARENT") ?: "/hbase")
            set(HConstants.ZOOKEEPER_QUORUM, getEnv("K2HB_HBASE_ZOOKEEPER_QUORUM") ?: "zookeeper")
            setInt("hbase.zookeeper.port", getEnv("K2HB_HBASE_ZOOKEEPER_PORT")?.toIntOrNull() ?: 2181)
            set(HConstants.HBASE_RPC_TIMEOUT_KEY, getEnv("K2HB_HBASE_RPC_TIMEOUT_MILLISECONDS") ?: "1200000")
            set(HConstants.HBASE_CLIENT_OPERATION_TIMEOUT, getEnv("K2HB_HBASE_OPERATION_TIMEOUT_MILLISECONDS") ?: "1800000")
            set(HConstants.HBASE_CLIENT_PAUSE, getEnv("K2HB_HBASE_PAUSE_MILLISECONDS") ?: "50")
            set(HConstants.HBASE_CLIENT_RETRIES_NUMBER, getEnv("K2HB_HBASE_RETRIES") ?: "50")
            set("hbase.client.keyvalue.maxsize", getEnv("K2HB_HBASE_KEYVALUE_MAX_SIZE") ?: "0")
        }

        val columnFamily = getEnv("K2HB_HBASE_COLUMN_FAMILY") ?: "cf"
        val columnQualifier = getEnv("K2HB_HBASE_COLUMN_QUALIFIER") ?: "record"
        val regionReplication: Int = getEnv("K2HB_HBASE_REGION_REPLICATION")?.toInt() ?: 3
        var DEFAULT_QUALIFIED_TABLE_PATTERN = """^\w+\.([-\w]+)\.([-.\w]+)$"""
        var qualifiedTablePattern = getEnv("K2HB_QUALIFIED_TABLE_PATTERN") ?: DEFAULT_QUALIFIED_TABLE_PATTERN
        val regionSplits = getEnv("K2HB_HBASE_REGION_SPLITS") ?: "2"
    }

    object Kafka {
        val consumerProps = Properties().apply {
            put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, getEnv("K2HB_KAFKA_BOOTSTRAP_SERVERS") ?: "kafka:9092")
            put(ConsumerConfig.GROUP_ID_CONFIG, getEnv("K2HB_KAFKA_CONSUMER_GROUP") ?: "test")
            put(ConsumerConfig.CLIENT_ID_CONFIG, "${InetAddress.getLocalHost().hostName}-${System.currentTimeMillis()}")
            addSslConfig(this)
            put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer::class.java)
            put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer::class.java)
            put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
            put(ConsumerConfig.METADATA_MAX_AGE_CONFIG, getEnv("K2HB_KAFKA_META_REFRESH_MS") ?: "10000")
            put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false)
            put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, getEnv("K2HB_KAFKA_MAX_POLL_RECORDS") ?: 500)
            val pollInterval = getEnv("K2HB_KAFKA_MAX_POLL_INTERVAL_MS")
            if (pollInterval != null) {
                put(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG, pollInterval.toInt())
            }
            put(ConsumerConfig.FETCH_MAX_BYTES_CONFIG, getEnv("K2HB_KAFKA_MAX_FETCH_BYTES") ?: 100_000_000)
            put(ConsumerConfig.MAX_PARTITION_FETCH_BYTES_CONFIG, getEnv("K2HB_KAFKA_MAX_PARTITION_FETCH_BYTES") ?: 100_000_000)
        }

        val producerProps = Properties().apply {
            put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, getEnv("K2HB_KAFKA_BOOTSTRAP_SERVERS") ?: "kafka:9092")
            addSslConfig(this)
            put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer::class.java)
            put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer::class.java)
            put(metaDataRefreshKey, getEnv("K2HB_KAFKA_META_REFRESH_MS") ?: "10000")
        }

        private fun addSslConfig(properties: Properties) {
            val insecure = getEnv("K2HB_KAFKA_INSECURE") ?: "true"
            if (insecure != "true") {
                properties.apply {
                    put("security.protocol", "SSL")
                    put(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, getEnv("K2HB_TRUSTSTORE_PATH"))
                    put(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, getEnv("K2HB_TRUSTSTORE_PASSWORD"))
                    put(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, getEnv("K2HB_KEYSTORE_PATH"))
                    put(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, getEnv("K2HB_KEYSTORE_PASSWORD"))
                    put(SslConfigs.SSL_KEY_PASSWORD_CONFIG, getEnv("K2HB_PRIVATE_KEY_PASSWORD"))
                }
            }
        }

        val pollTimeout: Duration = getEnv("K2HB_KAFKA_POLL_TIMEOUT")?.toDuration() ?: Duration.ofSeconds(3)
        val topicRegex = Regex(getEnv("K2HB_KAFKA_TOPIC_REGEX") ?: """^(db[.]{1}[-\w]+[.]{1}[-\w]+)$""")

        val topicExclusionRegexText = getEnv("K2HB_KAFKA_TOPIC_EXCLUSION_REGEX") ?: "NOT_SET"
        val topicExclusionRegex: Regex? = if(topicExclusionRegexText != "NOT_SET") Regex(topicExclusionRegexText) else null

        val dlqTopic = getEnv("K2HB_KAFKA_DLQ_TOPIC") ?: "test-dlq-topic"

    }

    object MetadataStore {
        val writeToMetadataStore = (getEnv("K2HB_WRITE_TO_METADATA_STORE") ?: "true").toBoolean()
        val metadataStoreTable = getEnv("K2HB_METADATA_STORE_TABLE") ?: "ucfs"
        val autoCommit = (getEnv("K2HB_METADATA_STORE_AUTO_COMMIT") ?: "false").toBoolean()

        private val useAwsSecretsString = getEnv("K2HB_USE_AWS_SECRETS") ?: "true"

        val isUsingAWS = useAwsSecretsString == "true"

        val properties = Properties().apply {
            put("user", getEnv("K2HB_RDS_USERNAME") ?: "k2hbwriter")
            put("rds.password.secret.name", getEnv("K2HB_RDS_PASSWORD_SECRET_NAME") ?: "password")
            put("database", getEnv("K2HB_RDS_DATABASE_NAME") ?: "metadatastore")
            put("rds.endpoint", getEnv("K2HB_RDS_ENDPOINT") ?: "127.0.0.1")
            put("rds.port", getEnv("K2HB_RDS_PORT") ?: "3306")
            put("use.aws.secrets", getEnv("K2HB_USE_AWS_SECRETS") ?: "true")

            if (isUsingAWS) {
                put("ssl_ca_path", getEnv("K2HB_RDS_CA_CERT_PATH") ?: "/certs/AmazonRootCA1.pem")
                put("ssl_ca", readFile(getProperty("ssl_ca_path")))
                put("ssl_verify_cert", true)
            }
        }
    }

    object SecretManager {
        val properties = Properties().apply {
            put("region", getEnv("SECRET_MANAGER_REGION") ?: dataworksRegion)
        }
    }

    object AwsS3 {
        private val maxS3Connections: Int = (getEnv("K2HB_AWS_S3_MAX_CONNECTIONS") ?: "2000").toInt()
        private val useLocalStack = (getEnv("K2HB_AWS_S3_USE_LOCALSTACK") ?: "false").toBoolean()
        private val region = getEnv("K2HB_AWS_S3_REGION") ?: dataworksRegion

        private const val localstackServiceEndPoint = "http://aws-s3:4566/"
        private const val localstackAccessKey = "AWS_ACCESS_KEY_ID"
        private const val localstackSecretKey = "AWS_SECRET_ACCESS_KEY"

        val s3: AmazonS3 by lazy {
            if (useLocalStack) {
                AmazonS3ClientBuilder.standard()
                    .withEndpointConfiguration(AwsClientBuilder.EndpointConfiguration(localstackServiceEndPoint, dataworksRegion))
                    .withClientConfiguration(ClientConfiguration().apply {
                        withProtocol(Protocol.HTTP)
                        maxConnections = maxS3Connections
                    })
                    .withCredentials(AWSStaticCredentialsProvider(BasicAWSCredentials(localstackAccessKey, localstackSecretKey)))
                    .withPathStyleAccessEnabled(true)
                    .disableChunkedEncoding()
                    .build()
            }
            else {
                AmazonS3ClientBuilder.standard()
                    .withCredentials(DefaultAWSCredentialsProviderChain())
                    .withRegion(region)
                    .withClientConfiguration(ClientConfiguration().apply {
                        maxConnections = maxS3Connections
                    })
                    .build()
            }
        }
    }

    object CorporateStorage {
        val archiveBucket = getEnv("K2HB_AWS_S3_ARCHIVE_BUCKET") ?: "ucarchive"
        val archiveDirectory = getEnv("K2HB_AWS_S3_ARCHIVE_DIRECTORY") ?: "ucdata_main"
    }

    object Manifest {
        val manifestBucket = getEnv("K2HB_AWS_S3_MANIFEST_BUCKET") ?: "manifests"
        val manifestDirectory = getEnv("K2HB_AWS_S3_MANIFEST_DIRECTORY") ?: "streaming"
        val writeManifests = (getEnv("K2HB_AWS_S3_WRITE_MANIFESTS") ?: "true").toBoolean()
    }

    object Retry {
        val maxAttempts: Int = getEnv("K2HB_RETRY_MAX_ATTEMPTS")?.toInt() ?: 3
        val initialBackoff: Long = getEnv("K2HB_RETRY_INITIAL_BACKOFF")?.toLong() ?: 10000
        val backoffMultiplier: Long = getEnv("K2HB_RETRY_BACKOFF_MULTIPLIER")?.toLong() ?: 2
    }

    fun getEnv(envVar: String): String? {
        val value = System.getenv(envVar)
        return if (value.isNullOrEmpty()) null else value
    }

    fun String.toDuration(): Duration = Duration.parse(this)
    fun readFile(fileName: String): String = File(fileName).readText(Charsets.UTF_8)

}
